pkg/operator/operator_test.go

```diff
@@ -0,0 +1,131 @@
+package operator
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"testing"
+	"time"
+
+	v1 "k8s.io/api/certificates/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+const dir = "_tmp"
+
+func TestEnsureCerts(t *testing.T) {
+	var (
+		timeout     = 3 * time.Second
+		ctx, cancel = context.WithTimeout(context.Background(), timeout)
+		csr         *v1.CertificateSigningRequest
+		fqdn        = fmt.Sprintf("system:node:%s.%s.svc", NameOperator, "")
+		certBytes   = []byte{1, 2, 3, 4}
+	)
+	t.Cleanup(cancel)
+
+	os.Mkdir(dir, 0777)
+	defer func() {
+		os.RemoveAll(dir)
+	}()
+
+	client := fake.NewSimpleClientset()
+	apiV1 := client.CertificatesV1().CertificateSigningRequests()
+	// Fire off a goroutine to mock asynchronous kube-apiserver.
+	go func() {
+		// Block until kube client has newly-created CSR.
+		if err := wait.Poll(50*time.Millisecond, timeout, func() (bool, error) {
+			var gerr error
+			// Ignore any errors coming from get API call as
+			csr, gerr = apiV1.Get(ctx, fqdn, metav1.GetOptions{})
+			if apierrors.IsNotFound(gerr) {
+				return false, nil
+			}
+			// Check if certificate has been approved by the signing function.
+			if len(csr.Status.Conditions) < 1 {
+				return false, nil
+			} else {
+				return csr.Status.Conditions[0].Type == v1.CertificateApproved, nil
+			}
+		}); err != nil {
+			t.Errorf("timeout waiting for CSR: %s", err)
+		}
+
+		// Mock the kube-apiserver issuing the certificate by writing toy data to status.
+		// This will unblock the keypair generation goroutine.
+		csr.Status.Certificate = certBytes
+		if _, err := apiV1.UpdateStatus(ctx, csr, metav1.UpdateOptions{}); err != nil {
+			t.Errorf("updating csr while issuing cert: %s", err)
+		}
+	}()
+
+	for _, tc := range []struct {
+		desc       string
+		opts       Options
+		expectCert string
+		expectKey  string
+		expectErr  bool
+	}{
+		{
+			desc:       "input key/cert",
+			opts:       Options{Key: "a2V5", Cert: "Y2VydA=="},
+			expectCert: "cert",
+			expectKey:  "key",
+			expectErr:  false,
+		},
+		{
+			desc:      "bad cert",
+			opts:      Options{Cert: "not a cert", Key: "not a key"},
+			expectErr: true,
+		},
+		{
+			desc:       "self generate keys/cert",
+			expectCert: "1,2,3,4",
+			expectErr:  false,
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			op := Operator{
+				opts:       tc.opts,
+				kubeClient: client,
+			}
+
+			_, err := op.ensureCerts(ctx, dir)
+			if err == nil == tc.expectErr {
```

Comment:
I got really confused syntactically initially. A more common way to write this would be with two conditions and dedicated errors (`err != nil && !tc.expectErr` and `err == nil && tc.expectErr`) or at least by bracing them (`err == nil) == tc.expectErr`).